File: data structures and algorithms/java/05_051.java
import java.util.*;
import java.util.stream.*;
import java.io.*;
import java.math.*;

public class Main {
    public static void main(String[] args) {
        // Main method code goes here
    }
    
    public class Node {
        public int key;
        public Node left;
        public Node right;
        
        public Node(int key) {
            this.key = key;
        }
    }
    
    public boolean isValidBST(Node root, long minVal, long maxVal) {
        if (root == null) {
            return true;
        }
        
        if (root.key <= minVal || root.key >= maxVal) {
            return false;
        }
        
        return (isValidBST(root.left, minVal, root.key) && 
                isValidBST(root.right, root.key, maxVal));
    }
    
    public boolean isValidBST(Node root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
}

File: data structures and algorithms/cs/05_051.cs
using System;

public class MainClass
{
   public static void Main(string[] args)
   {
      // Main method code goes here
   }

   public class Node
   {
      public int Key;
      public Node Left;
      public Node Right;

      public Node(int key)
      {
         this.Key = key;
      }
   }

   public bool IsValidBST(Node root, long minVal, long maxVal)
   {
      if (root == null)
      {
         return true;
      }

      if (root.Key <= minVal || root.Key >= maxVal)
      {
         return false;
      }

      return (IsValidBST(root.Left, minVal, root.Key) &&
              IsValidBST(root.Right, root.Key, maxVal));
   }

   public bool IsValidBST(Node root)
   {
      return IsValidBST(root, long.MinValue, long.MaxValue);
   }
}
